docs: per-object state metrics (#3104)
NVIDIA/bare-metal-manager-core: docs/design/per-object-state-metrics.md

# Per-Object State Progress Metrics

- **Issue:** [#2186](https://github.com/NVIDIA/infra-controller/issues/2186) · **Related:** [#191](https://github.com/NVIDIA/infra-controller/issues/191), [#2168](https://github.com/NVIDIA/infra-controller/pull/2168)
- **Status:** Proposal

## Problem

State-controller metrics are aggregate-only (`carbide_machines_per_state`,
`_per_state_above_sla`, transition counters, time-in-state histograms). At
100k-machine scale, SREs cannot answer via Prometheus alone:

1. **Which** machines are stuck beyond SLA (not just how many).
2. Which need **manual operator action**, and why.
3. What state a machine is in, **as a join key** (e.g. suppress "DPU not
   calling home" during DPU reprovision).
4. Per-machine **custom SLAs** (hardware gen, tenant) without per-fleet rules.
5. Slicing by **traits** (rack, SKU) and **associations** (host↔DPU,
   machine↔instance).

> 

Reviewer: Biggest reason to use `1.` for every list item is ease of maintenance. No need to renumber if you add/remove items. GitHub will render it as sequential, as will Fern for the public docs on docs.nvidia.com (although this isn't currently included in /fern/docs.yml, so that may not be a concern).

Also, side note, why is everyone hard-breaking their lines in this repo? It's such a hassle to edit text this way.

```suggestion
1. **Which** machines are stuck beyond SLA (not just how many).
1. Which need **manual operator action**, and why.
1. What state a machine is in, **as a join key** (for example, suppress
   "DPU not calling home" during DPU reprovision).
1. Per-machine **custom SLAs** (hardware gen, tenant) without per-fleet rules.
1. Slicing by **traits** (rack, SKU) and **associations** (host↔DPU,
   machine↔instance).
```

- `state_sla(state, object_state)` receives the full object state, so
  per-hardware-gen/tenant SLAs are a policy change in that function only.
- PR #2168's `PerObjectMetricsRegistry` (`health-metrics/src/per_object.rs`)
  was built to generalize across object types and is **reused directly**:
  shared registry keyed by `(object_type, object_id)`, observable gauges,
  lazy stale eviction, config-gated.
- Manual-intervention signals exist per-object:
  `StateHandlerError::ManualInterventionRequired` and terminal
  `Failed { FailureDetails }` states.

## Design

### Dedicated endpoint

Per-object series live on a second Prometheus registry/meter, served on its
own listener:

```toml
[observability.per_object_state_metrics]
enabled = true                    # default: false
listen_address = "0.0.0.0:9091"
object_types = ["machine", "switch", "power_shelf", "rack"]  # default: all
```

This lets operators scrape it slower (60–120s — the series change only on
state transitions), route it to a different tenant/retention, or skip it —
without touching the existing `/metrics`. `api-core` already owns its metrics
handler; adding a second `(registry, meter, listener)` in

> `run.rs::create_metrics()` is mechanical.

Author: This is fine, we will have separate endoint for such metrics, Freshness are not issue for them

These two triggers are the complete v1 set — precision here matters more than
recall; further triggers (e.g. substate `Failed` variants with exhausted
retries) are added only after operational experience. `TimeInStateAboveSla`
is deliberately **excluded**: SLA breach is a symptom, fully expressible from
the two metrics above; mixing it in makes this set noisy during mass-slowness
incidents, exactly when it must be precise.

#### `carbide_object_info` (value 1) — stable traits

```text
carbide_object_info{object_type="machine",object_id="fm100...",rack_id="rack-12",sku="GB200-NVL72-A",vendor="Supermicro",model="SYS-421GE"} 1
```

Standard `_info` join pattern (cf. `kube_node_info`). Traits live *only* here —
on state series they'd multiply cardinality and churn on correction. Caveat:
vendor/model aren't first-class `machines` columns (they come from exploration
reports), so those labels are best-effort `""` when unknown.

#### Association info metrics (value 1, one series per pair)

```text
carbide_machine_dpu_info{machine_id="fm100...",dpu_id="fmdpu01..."} 1
carbide_machine_instance_info{machine_id="fm100...",instance_id="inst-42...",tenant_org="acme"} 1

> ```

Author: We would need tenants to answer specific question. So we assume we can break this rule just for TenantId which should be very limited in scope.

> ```

Author: @fengzhou-nv can maybe we drop this requirement for tenant_id? Or it absolutely required?

> ```

Reviewer 2: I think we need to have `tenant_id` or something similar. This allows us to answer following questions directly via proemtheus such as

- what instances are owned by a particular tenant, or what machines
- which of the machines that belong to a tenant are out of sla/having issues
- what impact of <something> has on <xyz> tenant (in post-incident analysis)

One series per *relationship* resolves the multi-DPU-host concern from the
issue comments: a 4-DPU host is 4 association series while its state series
stays one — no `blocking_dpu_id` label on it either; the host reports the
least-progressed DPU's substate (existing `metric_state_names` behavior), and
since DPUs are machines with their own state series, the association join

> identifies the blocking DPU.

Reviewer: I think I'd rather not see "concern from the issue comments" here. Maybe something like this?

```suggestion
One series per *relationship* resolves a concern about multiple DPU hosts:
a 4-DPU host is 4 association series, while its state series stays one. There
is also no `blocking_dpu_id` label on it; the host reports the
least-progressed DPU's substate (matching the `metric_state_names`
behavior), and since DPUs are machines with their own state series, the
association join identifies the blocking DPU.
```

Instances get no state series of their own: they are machine substates
(`Assigned { instance_state }`), not separate state-controller objects, so
they surface as `state="assigned", substate=<instance state>` on the machine

> plus `carbide_machine_instance_info` for the id mapping — a synthetic

Reviewer: ```suggestion
plus `carbide_machine_instance_info` for the ID mapping — a synthetic
```

per-instance series would double cardinality for no new information.

### Cardinality budget (100k machines + ~5k other objects)

| Metric | Series |
|---|---|
| `state_entered_timestamp` | ~105k (1/object) |
| `state_sla` | ≪ fleet (SLA states only) |
| `manual_intervention_required` | fleet-in-trouble (~5k at the issue's 5%) |
| `object_info` | ~105k |
| associations | ~100k + #instances |
| **Total** | **~400–500k slow-moving series**, opt-in, on their own endpoint |

Churn is bounded by transition rate, observable in advance via the existing
`_state_entered_total` counters.

## Queries (issue use cases)

**Stuck beyond per-object SLA** (warning; critical = `2 *`):

> ```promql

Reviewer: I can pretty much guarantee this won't get source-highlighted (here is [Fern's list of supported languages](https://buildwithfern.com/learn/docs/writing-content/components/code-blocks#usage)), but it's not going to break, either. If you want to try and get some useful highlighting, you might try seeing what happens with `sql` for these.)

(time() - carbide_object_state_entered_timestamp_seconds)
  > on(object_type, object_id, state, substate) group_left()
    carbide_object_state_sla_seconds
```

**Manual-intervention ratio and triage breakdown:**

```promql
count(carbide_object_manual_intervention_required{object_type="machine"})
  / scalar(carbide_machines_total) > 0.05

count by (reason, rack_id) (
  carbide_object_manual_intervention_required{object_type="machine"}
  * on(object_id) group_left(rack_id, sku) carbide_object_info)
```

**Suppress DPU-not-calling-home during reprovision** (join use case;
`DPUReprovision` maps to `state="reprovisioning"` today):

```promql
(time() - carbide_forge_dpu_agent_last_call > 900)
and on(dpu_id) label_replace(carbide_machine_dpu_info, "dpu_id", "$1", "dpu_id", "(.*)")
unless on(machine_id) label_replace(
  carbide_object_state_entered_timestamp_seconds{state="reprovisioning"},
  "machine_id", "$1", "object_id", "(.*)")
```

> **Stuck-and-unhealthy — is it hardware?** (join with the #2168 metric):

Reviewer: Uh... GitHub issue numbers directly in docs? I'd prefer this at least be a real URL to the issue, and even better, a one-liner explanation from said issue (with a link).

Author: This is design doc, i do not think they are rendered in docs? At least i think they should not be. If there is wrong place, where we put design (RFC)?

  Prometheus needs "where is it now, since when" — provided above.
- **`expected_next_state` per object.** State machines are imperative handler
  code; the next state depends on runtime conditions. A *static* per-`(state,
  substate)` "typical next state" info metric could be added later.
- **Dependency-lag metrics** (network config sync, IB sync) — need their own
  instrumentation design. Reserved shape:
  `carbide_object_dependency_lag_seconds{object_id, dependency="network_config_sync"|...}`.
- **Hardware-cause detail** — covered by joining the existing per-object
  health classification metric (query above); probe-level detail stays on
  aggregate `_unhealthy_by_probe_id_count`.

## Implementation

**Reuse `PerObjectMetricsRegistry`** (`health-metrics/src/per_object.rs`).
It was built for exactly this generalization — a single shared registry keyed
by `(object_type, object_id)` "so the metric name stays stable as
observability generalizes across object types" (its module doc) — and already
provides everything the new metrics need: shared entry map, gauge callbacks
over live entries, `hold_period` eviction, replace/remove-on-record

> semantics. We extend it rather than adding a sibling:

Reviewer: ```suggestion
over live entries, `hold_period` eviction, and replace/remove-on-record
semantics.

We extend the `PerObjectMetricsRegistry`, rather than adding a sibling:
```

- Generalize the entry payload from classification-only to per-metric series:
  a `gauge(name, description)` handle API where writers `set`/`set_all`/
  `clear` an object's series for that metric (the existing classification
  gauge becomes the first handle, keeping its config gating and behavior).
- Let each handle register on a caller-supplied `Meter`, so the new state
  gauges land on the per-object endpoint while the existing
  `carbide_object_unhealthy_by_classification_count` stays on the main
  endpoint (moving it is a scrape-config-visible change; do it as a later,

>   separately announced step).

Reviewer: This feels like an internal note. If there's a plan to make the change, maybe link to the issue for tracking? And if there's no plan yet, I'd be inclined to remove.

Author: It is more about RFC/Design, not docs, so i think it is justified here

**Feed point: the generic processor** (`processor.rs`), which already holds

> object id, (transitioned) state, `time_in_state`, `StateSla`, and handler

Reviewer: ```suggestion
object ID, (transitioned) state, `time_in_state`, `StateSla`, and handler
```

error per iteration — one `registry.record_state(...)` call there gives
**every** state controller per-object metrics with zero per-controller code.
Wiring mirrors #2168: constructed in `setup.rs`, registered on the per-object
meter, threaded via the controller builder. `carbide_object_info` and
associations are recorded from the machine-controller handler, which already

> loads rack/SKU/DPU/instance data next to the existing per-object health call.

Reviewer: Again, add a real URL at least, but probably better to add real description (and also possibly a link to the issue/PR).
```suggestion
This wiring mirrors #2168: constructed in `setup.rs`, registered on the
per-object meter, threaded via the controller builder. `carbide_object_info`
and associations are recorded from the machine-controller handler, which
already loads rack/SKU/DPU/instance data next to the existing per-object
health call.
```

Author: Added link, i find it is cleaner to look at source, it have more context than just explaining it in the RFC

> 

Reviewer: Linter is complaining about multiple blank lines at EOF, but I'm only seeing one. Weird. If there's more than one, make it exactly one. :)
